File: Python/20201202.py
from numpy import loadtxt


def parse_pass_policy(password_line):
    policy, password = password_line.split(":")
    policy_num, policy_letter = policy.split()
    policy_low, policy_high = policy_num.split("-")

    password = password.strip()

    policy_low = int(policy_low)
    policy_high = int(policy_high)

    return policy_low, policy_high, policy_letter, password


def check_valid_password(password_line):
    low, high, letter, password = parse_pass_policy(password_line)

    letter_count = password.count(letter)

    if (letter_count >= low) and (letter_count <= high):
        return 1
    else:
        return 0


def check_pass_part_2(password_line):
    first, second, letter, password = parse_pass_policy(password_line)

    if password[first-1] == letter:
        if password[second-1] == letter:
            return 0
        else:
            return 1
    else:
        if password[second-1] == letter:
            return 1
        else:
            return 0


if __name__ == "__main__":
    lines = open("../input/day2.txt", "r")
    lines = lines.readlines()

    valid_passwords = 0
    valid_passwords_2 = 0

    for password in lines:
        valid_passwords += check_valid_password(password.rstrip())
        valid_passwords_2 += check_pass_part_2(password.rstrip())

    print(valid_passwords)
    print(valid_passwords_2)

File: Go/20201202.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

func readLines(filePath string) []string {
	f, err := os.Open(filePath)

	if err != nil {
		log.Fatal(err)
	}

	defer f.Close()

	scanner := bufio.NewScanner(f)

	passwords := []string{}

	for scanner.Scan() {
		passwords = append(passwords, scanner.Text())
	}

	if err := scanner.Err(); err != nil {
		log.Fatal(err)
	}

	return passwords
}

func part1(passwordList []string) int {
	validPasswords := 0

	for _, passwordFull := range passwordList {
		policySlice := strings.Split(passwordFull, " ")

		policy := policySlice[0]
		policies := strings.Split(policy, "-")
		lower, _ := strconv.Atoi(policies[0])
		upper, _ := strconv.Atoi(policies[1])

		letter := strings.Split(policySlice[1], "")[0]

		password := policySlice[2]

		letterCount := strings.Count(password, letter)

		if lower <= letterCount && letterCount <= upper {
			validPasswords++
		}
	}
	return validPasswords
}

func part2(passwordList []string) int {
	validPasswords := 0

	for _, passwordFull := range passwordList {
		policySlice := strings.Split(passwordFull, " ")
		policy := policySlice[0]
		policies := strings.Split(policy, "-")
		first, _ := strconv.Atoi(policies[0])
		second, _ := strconv.Atoi(policies[1])

		letter := strings.Split(policySlice[1], "")[0]
		password := strings.Split(policySlice[2], "")

		firstCheck := password[first-1] == letter
		secondCheck := password[second-1] == letter

		if firstCheck {
			if !secondCheck {
				validPasswords++
			}
		}
		if secondCheck {
			if !firstCheck {
				validPasswords++
			}
		}
	}
	return validPasswords
}

func main() {
	passwords := readLines("../input/day2.txt")
	part1Count := part1(passwords)
	fmt.Println("Part 1: ", part1Count)

	part2Count := part2(passwords)
	fmt.Println("Part 2: ", part2Count)
}
